filebeat/harvester/log.go

```diff
@@ -79,45 +107,46 @@ func (h *Harvester) Harvest() {
 	// TODO: NewLineReader uses additional buffering to deal with encoding and testing
 	//       for new lines in input stream. Simple 8-bit based encodings, or plain
 	//       don't require 'complicated' logic.
-	timedIn := newTimedReader(h.file)
-	reader, err := encoding.NewLineReader(timedIn, enc, h.Config.BufferSize)
+	config := h.Config
+	readerConfig := logFileReaderConfig{
+		forceClose:         config.ForceCloseFiles,
+		maxInactive:        h.ProspectorConfig.IgnoreOlderDuration,
+		backoffDuration:    config.BackoffDuration,
+		maxBackoffDuration: config.MaxBackoffDuration,
+		backoffFactor:      config.BackoffFactor,
+	}
+	reader, err := createLineReader(h.file, enc, config.BufferSize, readerConfig, config.Multiline)
 	if err != nil {
 		logp.Err("Stop Harvesting. Unexpected encoding line reader error: %s", err)
 		return
 	}
 
-	// XXX: lastReadTime handling last time a full line was read only?
-	//      timedReader provides timestamp some bytes have actually been read from file
-	lastReadTime := time.Now()
-
 	for {
 		// Partial lines return error and are only read on completion
-		text, bytesRead, err := readLine(reader, &timedIn.lastReadTime)
-
+		text, bytesRead, err := readLine(reader)
 		if err != nil {
-
-			// In case of err = io.EOF returns nil
-			err = h.handleReadlineError(lastReadTime, err)
-
-			// Return in case of error which leads to stopping harvester and closing file
-			if err != nil {
-				logp.Info("Read line error: %s", err)
-				return
+			if err == errFileTruncate {
+				seeker, ok := h.file.(io.Seeker)
+				if !ok {
+					logp.Err("can not seek source")
+					return
+				}
+
+				logp.Debug("harvester", "File was truncated. Begin reading file from offset 0: %s", h.Path)
+
+				h.Offset = 0
+				seeker.Seek(h.Offset, os.SEEK_SET)
+				continue
 			}
 
-			continue
+			logp.Info("Read line error: %s", err)
+			return
 		}
 
-		lastReadTime = time.Now()
-
-		// Reset Backoff
-		h.backoff = h.Config.BackoffDuration
-
 		if h.shouldExportLine(text) {
-
 			// Sends text to spooler
 			event := &input.FileEvent{
-				ReadTime:     lastReadTime,
+				ReadTime:     time.Now(),
```

Comment:
lastReadTime was set to time.Now() right before creating the event: https://github.com/elastic/beats/blob/master/filebeat/harvester/log.go#L111

I just removed the variable. We can update the lineReader interface (or create another interface for last line-reading stage) to return timestamps and include the timestamp in the state-machine.